### ISSUE TITLE
Kotlin classes incompatible with RedissonLiveObjectService 

### ISSUE THREAD

**Issue by michaelcordero** [redisson/redisson#2977]
**Expected behavior**
User should be able to register arbitrary Kotlin classes.

**Actual behavior**
ByteBuddy throws `new IllegalArgumentException("Cannot subclass primitive, array or final types: " + superType);` on normal class declarations. This is because Kotlin's classes are `final` by default. However, to allow sub-classes, even after the `open` keyword is applied, and `constructor` keyword declared. ByteBuddy is still failing, with: `java.lang.IllegalArgumentException: Can't find default constructor for class Customer$ByteBuddy$sa6Yvns1`.

**Steps to reproduce or test case**
```
import kotlinx.coroutines.*
import org.redisson.Redisson
import org.redisson.api.RLiveObjectService
import org.redisson.api.RedissonClient
import org.redisson.api.annotation.REntity
import org.redisson.api.annotation.RId
import org.redisson.api.annotation.RIndex

@REntity
open class Customer constructor(@RId val id: Int, @RIndex var name: String?, @RIndex var city: String?)

@ExperimentalCoroutinesApi
@InternalCoroutinesApi
fun main(): Unit {
        val client: RedissonClient = Redisson.create() // connects to 127.0.0.1:6379 by default
        val service: RLiveObjectService = client.liveObjectService
        service.registerClass(Customer::class.java)
        val customers: List<Customer> = listOf(
            Customer(1, "angela", "LA"),
            Customer(2, "svetlana", "moscow"),
            Customer(3, "juliana", "kiev")
        )
        customers.forEach { service.persist(it) }
}
```

**Redis version**
6.0.1
**Redisson version**
3.13.3

**Redisson configuration**
default

**Comment by mrniko** [comment]
You need to add default constructor

**Comment by michaelcordero** [comment]
The default constructor is erasing the data.

```
import kotlinx.coroutines.*
import org.redisson.Redisson
import org.redisson.api.RLiveObjectService
import org.redisson.api.RedissonClient
import org.redisson.api.annotation.REntity
import org.redisson.api.annotation.RId
import org.redisson.api.annotation.RIndex

@REntity
open class Customer constructor(@RId var id: Int, @RIndex var name: String?, @RIndex var city: String?) {
    constructor(): this(0,"","")
}

@ExperimentalCoroutinesApi
@InternalCoroutinesApi
fun main(): Unit {
    val client: RedissonClient = Redisson.create() // connects to 127.0.0.1:6379 by default
    val service: RLiveObjectService = client.liveObjectService
    service.registerClass(Customer::class.java)
    val customers: List<Customer> = listOf(
        Customer(1, "angela", "LA"),
        Customer(2, "svetlana", "moscow"),
        Customer(3, "juliana", "kiev")
    )
    customers.forEach { service.persist(it) }
    println("Fetching Customer #2: ")
    val cust: Customer = service.get(Customer::class.java,2)
    println(cust.toString())
}
```

**Comment by hanzo2001** [comment]
I have a feeling that **LiveObjectService** is not ready for Kotlin.

I used the above example from @michaelcordero and expanded it to show that it really does not matter whether you have or not a default constructor. The fields get reset and I've confirmed that the data that ends up being stored in Redis is actually empty. The object is there but none of its fields are populated.

```kt
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.redisson.Redisson
import org.redisson.api.RLiveObjectService
import org.redisson.api.RedissonClient
import org.redisson.api.annotation.REntity
import org.redisson.api.annotation.RId
import org.redisson.api.annotation.RIndex
import org.redisson.config.Config

/**
 * I thought the issue was that the proxy getters and setters were not being called on the live object...
 */
@REntity
open class C1() {

    @RId    private var _id: Int = 0
    @RIndex private var _name: String = ""

    var id
        get() = _id
        set(value) { this._id = value }
    var name
        get() = _name
        set(value) { this._name = value }

    constructor(
        id: Int,
        name: String,
    ) : this() {
        this.id = id
        this.name = name
    }

    override fun toString() = """{"id":$id,"name":"$name"}"""
}

/**
 * Maybe the property accessors are not the issue...
 */
@REntity
open class C2() {

    @RId    var id: Int = 0
    @RIndex var name: String = ""

    constructor(
        id: Int,
        name: String,
    ) : this() {
        this.id = id
        this.name = name
    }

    override fun toString() = """{"id":$id,"name":"$name"}"""
}

/**
 * data classes are guaranteed to fail because they are closed and Redisson needs to extend on them to create the proxy
 */
@REntity
data class C3(
    @RId    var id: Int = 0,
    @RIndex var name: String = "",
) {
    override fun toString() = """{"id":$id,"name":"$name"}"""
}

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class ByeTest {

    val config = Config()
    val client: RedissonClient
    val service: RLiveObjectService

    init {
        config
            .useClusterServers()
            .apply {
                (1..6).forEach { this.addNodeAddress("redis://127.0.0.1:3000$it") }
            }
        this.client = Redisson.create(config)
        this.service = client.liveObjectService
    }

    @Test
    fun a() {
        val c = C1::class.java
        service.registerClass(c)
        val customers = listOf(
            C1(1, "angela"),
            C1(2, "svetlana"),
            C1(3, "juliana"),
        )
        customers.mapNotNull { service.get(c, it.id) }.forEach(service::delete)
        customers.forEach { service.persist(it) }
        customers
            .onEach { println(it.toString()) }
            .map { it.id }
            .map { service.get(c, it) }
            .onEach { println(it.toString()) }
            .forEach(service::delete)
    }

    @Test
    fun b() {
        val c = C2::class.java
        service.registerClass(c)
        val customers = listOf(
            C1(1, "angela"),
            C1(2, "svetlana"),
            C1(3, "juliana"),
        )
        customers.mapNotNull { service.get(c, it.id) }.forEach(service::delete)
        customers.forEach { service.persist(it) }
        customers
            .onEach { println(it.toString()) }
            .map { it.id }
            .map { service.get(c, it) }
            .onEach { println(it?.toString() ?: "Nothing!") }
            .filterNotNull()
            .forEach(service::delete)
    }

    @Test
    fun c() {
        val c = C3::class.java
        service.registerClass(c)
        val customers = listOf(
            C1(1, "angela"),
            C1(2, "svetlana"),
            C1(3, "juliana"),
        )
        customers.mapNotNull { service.get(c, it.id) }.forEach(service::delete)
        customers.forEach { service.persist(it) }
        customers
            .onEach { println(it.toString()) }
            .map { it.id }
            .map { service.get(c, it) }
            .onEach { println(it?.toString() ?: "Nothing!") }
            .filterNotNull()
            .forEach(service::delete)
    }

}
```

### Test a
```
{"id":1,"name":"angela"}
{"id":2,"name":"svetlana"}
{"id":3,"name":"juliana"}
{"id":0,"name":""}
{"id":0,"name":""}
{"id":0,"name":""}
```

### Test b
```
{"id":1,"name":"angela"}
{"id":2,"name":"svetlana"}
{"id":3,"name":"juliana"}
Nothing!
Nothing!
Nothing!
```

### Test c
```
java.lang.IllegalArgumentException: Cannot subclass primitive, array or final types: class local.test.C3
```

I have no idea what it would take to make Redisson understand Kotlin property accessors and I hope that this example in fact helps future development.

**Comment by hanzo2001** [comment]
I am really interested in this feature. If possible, and as long as it is not an encumbrance, I would like to help speed up the completion of the 3.15.1 milestone by at least contributing to this specific point. Is there any way to get in touch or get started? Is there a "How to contribute" guide?

**Comment by hanzo2001** [comment]
I have read your contribution guideline which I hadn't seen before (https://github.com/redisson/redisson/blob/redisson-3.15.0/.github/CONTRIBUTING.md ... I hadn't seen the link write below the comment box x_x)

I started looking into this issue and I now realize that the actual proxy magic is coming from an external dependency called ByteBuddy. However, the proxy is built following the implementation at

```java
public class RedissonLiveObjectService implements RLiveObjectService {
    private <T> Class<? extends T> createProxy(Class<T> entityClass, CommandAsyncExecutor commandExecutor);
}
```

### Note 1

I have seen that when the cached proxy is used to instantiate a live proxy instance for a Kotlin class, the instantiation stores something that looks more like an exception (when debugging with Intellij, you can actually see the message: `Unable to evaluate the expression Method threw 'java.lang.NullPointerException' exception.`). I don't know if this is expected or not.

### Note 2

Maybe a completely different implementation of `::createProxy` is required to handle classes that come from kotlin. Property accessors are very different and reflection doesn't help

### Note 3

How would one write a test to see this issue resolved? Should there be a module for some Kotlin based resources?

**Comment by mrniko** [comment]
@hanzo2001 

How can we check if class was written on Kotlin to disable type validation?

> How would one write a test to see this issue resolved? Should there be a module for some Kotlin based resources?

Can we test Kotlin using only kotlin specific dependencies on test level?

**Comment by mrniko** [comment]
I applied necessary changes. Can you try version attached?

[redisson-3.15.1-SNAPSHOT.jar.zip](https://github.com/redisson/redisson/files/5949739/redisson-3.15.1-SNAPSHOT.jar.zip)

**Comment by hanzo2001** [comment]
I recreated the test bed and provided all the dependencies to my Kotlin project. The outcome is still the same

```
Test: given private backing fields when using field accessors then just work ****************
{"id":1,"name":"angela"}
{"id":2,"name":"svetlana"}
{"id":3,"name":"juliana"}
{"id":0,"name":""}
{"id":0,"name":""}
{"id":0,"name":""}

Test: given expected kotlin code then just work ****************
{"id":1,"name":"angela"}
{"id":2,"name":"svetlana"}
{"id":3,"name":"juliana"}
{"id":0,"name":""}
{"id":0,"name":""}
{"id":0,"name":""}

public final void local.test.ByeTest.given a data class then maybe work but not as a subclass() is @Disabled

Test: given some backing fields when using bean-like methods then just work ****************
{"id":1,"name":"angela"}
{"id":2,"name":"svetlana"}
{"id":3,"name":"juliana"}
{"id":0,"name":""}
{"id":0,"name":""}
{"id":0,"name":""}
```

I updated the test class file as follows

```kt
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInfo
import org.junit.jupiter.api.TestInstance
import org.redisson.Redisson
import org.redisson.api.RLiveObjectService
import org.redisson.api.RedissonClient
import org.redisson.api.annotation.REntity
import org.redisson.api.annotation.RFieldAccessor
import org.redisson.api.annotation.RId
import org.redisson.api.annotation.RIndex
import org.redisson.config.Config

@REntity
open class C1() {

    @RId    private var _id: Int = 0
    @RIndex private var _name: String = ""

    var id
        @RFieldAccessor
        get() = _id
        @RFieldAccessor
        set(value) { this._id = value }
    var name
        @RFieldAccessor
        get() = _name
        @RFieldAccessor
        set(value) { this._name = value }

    constructor(
        id: Int,
        name: String,
    ) : this() {
        this.id = id
        this.name = name
    }

    override fun toString() = """{"id":$id,"name":"$name"}"""
}

@REntity
open class C2() {

    @RId    var id: Int = 0
    @RIndex var name: String = ""

    constructor(
        id: Int,
        name: String,
    ) : this() {
        this.id = id
        this.name = name
    }

    override fun toString() = """{"id":$id,"name":"$name"}"""
}
@REntity
data class C3(
    @RId    var id: Int = 0,
    @RIndex var name: String = "",
) {
    override fun toString() = """{"id":$id,"name":"$name"}"""
}
@REntity
open class C4() {

    @RId    var id: Int = 0
    private var _name: String = ""
    @RFieldAccessor fun getName() = _name
    @RFieldAccessor fun setName(value: String) { this._name = value }

    constructor(
        id: Int,
        name: String,
    ) : this() {
        this.id = id
        this._name = name
    }

    override fun toString() = """{"id":$id,"name":"$_name"}"""
}

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class ByeTest {

    val config = Config()
    val client: RedissonClient
    val service: RLiveObjectService

    init {
        config
            .useClusterServers()
            .apply {
                (1..6).forEach { this.addNodeAddress("redis://127.0.0.1:3000$it") }
            }
        this.client = Redisson.create(config)
        this.service = client.liveObjectService
    }

    @Test
    fun `given private backing fields when using field accessors then just work`() {
        val c = C1::class.java
        service.registerClass(c)
        val customers = listOf(
            C1(1, "angela"),
            C1(2, "svetlana"),
            C1(3, "juliana"),
        )
        customers.mapNotNull { service.get(c, it.id) }.forEach(service::delete)
        customers.forEach {
            service.persist(it)
        }
        customers
            .onEach { println(it.toString()) }
            .map { it.id }
            .map { service.get(c, it) }
            .onEach { println(it.toString()) }
            .forEach(service::delete)
    }

    @Test
    fun `given expected kotlin code then just work`() {
        val c = C2::class.java
        service.registerClass(c)
        val customers = listOf(
            C2(1, "angela"),
            C2(2, "svetlana"),
            C2(3, "juliana"),
        )
        customers.mapNotNull { service.get(c, it.id) }.forEach(service::delete)
        customers.forEach { service.persist(it) }
        customers
            .onEach { println(it.toString()) }
            .map { it.id }
            .map { service.get(c, it) }
            .onEach { println(it?.toString() ?: "Nothing!") }
            .filterNotNull()
            .forEach(service::delete)
    }

    @Test
    @Disabled
    fun `given a data class then maybe work but not as a subclass`() {
        val c = C3::class.java
        service.registerClass(c)
        val customers = listOf(
            C3(1, "angela"),
            C3(2, "svetlana"),
            C3(3, "juliana"),
        )
        customers.mapNotNull { service.get(c, it.id) }.forEach(service::delete)
        customers.forEach { service.persist(it) }
        customers
            .onEach { println(it.toString()) }
            .map { it.id }
            .map { service.get(c, it) }
            .onEach { println(it?.toString() ?: "Nothing!") }
            .filterNotNull()
            .forEach(service::delete)
    }

    @Test
    fun `given some backing fields when using bean-like methods then just work`() {
        val c = C4::class.java
        service.registerClass(c)
        val customers = listOf(
            C4(1, "angela"),
            C4(2, "svetlana"),
            C4(3, "juliana"),
        )
        customers
            .mapNotNull {
                service.get(c, it.id)
            }
            .forEach {
                service.delete(it)
            }
        customers.forEach {
            service.persist(it)
        }
        customers
            .onEach { println(it.toString()) }
            .map { it.id }
            .map { service.get(c, it) }
            .onEach { println(it?.toString() ?: "Nothing!") }
            .filterNotNull()
            .forEach(service::delete)
    }

    @BeforeEach
    fun before(info: TestInfo) {
        println("Test: ${info.testMethod.get().name} ****************")
    }
}
```

I am not making any assertions, I am just checking output.

**Comment by mrniko** [comment]
Seems disable type validation isn't enough

**Comment by mrniko** [comment]
Here is the code I used:

```java
        ByteBuddy bb = new ByteBuddy();
        Class KOTLIN_METADATA = Class.forName("kotlin.Metadata");
        if (KOTLIN_METADATA != null && entityClass.getDeclaredAnnotation(KOTLIN_METADATA) != null) {
            bb.with(TypeValidation.DISABLED);
        }
        DynamicType.Builder<T> builder = bb.subclass(entityClass);
```

**Comment by hanzo2001** [comment]
I think the issue lies more in ByteBuddy's `ElementMatchers` and the proxy interception. I am trying to write a snippet in kotlin code to see if I can emulate some interceptors that actually work.

**Comment by hanzo2001** [comment]
I believe that this issue is completely contained inside the `RedissonLiveObjectService::createProxy` method and the methodology behind the configuration of the proxies. When I run the tests, the `persist` invocation persists what seems to be an empty object. The ID is correct but what I fetch back is obviously empty.

I am still experimenting with BB but this is quite other-worldly for me. I attached a snapshot of the state of my local RedisCluster while running the tests in debug mode. You can see that the key was correctly generated but there is nothing else inside. When I am working in Java, it should be possible to see the object's fields there.

![image](https://user-images.githubusercontent.com/2978880/107354989-c79dab80-6ac6-11eb-9a44-227ec6453021.png)

**Comment by hanzo2001** [comment]
I may be completely wrong but I am trying to work around both the matching and the interception inside the `createProxy`

```java
private <T> Class<? extends T> createProxy(Class<T> entityClass, CommandAsyncExecutor commandExecutor) {
    // ...
    Class<? extends T> proxied = builder
        // ... other implementations and intercepts ...
        // target a pub/prot get/set that is not one of the ones defined previously
        .method(
            not(isDeclaredBy(Object.class))
                .and(not(isDeclaredBy(RLiveObject.class)))
                .and(not(isDeclaredBy(RExpirable.class)))
                .and(not(isDeclaredBy(RExpirableAsync.class)))
                .and(not(isDeclaredBy(RObject.class)))
                .and(not(isDeclaredBy(RObjectAsync.class)))
                .and(not(isDeclaredBy(ConcurrentMap.class)))
                .and(not(isDeclaredBy(Map.class)))
                .and(isGetter().or(isSetter()))
                .and(isPublic().or(isProtected()))
        )
        .intercept(MethodDelegation.to(new AccessorInterceptor(commandExecutor, connectionManager)))
        .make()
        .load(entityClass.getClassLoader(), ClassLoadingStrategy.Default.WRAPPER)
        .getLoaded();
    return proxied;
}
```

I have a suspicion that the requirements of `isGetter` and `isSetter` may be causing the issue. I don't know if this is matching anything inside Kotlin.
Also, the delegation to `AccessorInterceptor` would require a review to make sure that it is targeting the correct fields.

**Comment by hanzo2001** [comment]
I have set up on a fork what seems to be a crude initial attempt at testing kotlin classes. I don't know how to share this with you but it is here `https://github.com/hanzo2001/redisson/tree/hanzo2001/fix/kotlin-in-liveobjectservice`

I am updating my branch (`hanzo2001/fix/kotlin-in-liveobjectservice`) with whatever I can think of.

My current attempt is to manually mark fields with a new annotation class `RKotlinMember`. I want to see this work before I can automate the process with ByteBuddy's more generic method selectors.

The test class is nothing more than a vulgar copy of `RedissonLiveObjectServiceTest` with some modifications. I moved many test entities outside of the class and I have set them up as kotlin classes. There are still a few models that I have not transferred yet but all in due time

**Comment by mrniko** [comment]
> I have a suspicion that the requirements of isGetter and isSetter may be causing the issue. I don't know if this is matching anything inside Kotlin.

It could be a problem since AccessorInterceptor need them to handle read/write access

**Comment by mrniko** [comment]
>My current attempt is to manually mark fields with a new annotation class RKotlinMember

You can use `RFieldAccessor` instead. Anyway you need a method so BB can intercept it.

**Comment by hanzo2001** [comment]
I am an idiot. The devil is in the details. I only fell off the tree when I looked into the method matchers and realized that the settings require that the methods be volatile and not final which in kotlin means that the must all be defined as `open`.

Another little detail is that all fields must be declared as `var`.

I found out that Kotlin's `@JvmOverloads` annotation for constructors with default parameters does not work due to an invisible arguments exception that pops up from ByteBuddy. This means that we cannot use Kotlin's easy magics on constructors.

The **Redisson** annotations are perfectly functional, no need for any extra magic. I also tested it while keeping type validation enabled and the test class runs clean.

I am so sorry to have wasted your time and for all the inconveniences I may have caused you.

![image](https://user-images.githubusercontent.com/2978880/107699927-aa2b3600-6cae-11eb-94f5-cd05b6ee82ee.png)

I will be cleaning up my branch and sending the changes later. All I'll be providing is a replica of `RedissonLiveObjectServiceTest` with a copy of all the required auxiliary test classes.

I added a little proposal for configuring the port on the `RedisRunner` that is read from a system var for those of us who wish to keep it fixed when debugging the contents inside the Redis instance.
